File: Army(c++)/Warriors/Necromancer.cpp
#include "Necromancer.hpp"
#include "../State/State.hpp"
#include "../Magic/SpellBook.hpp"

Necromancer::Necromancer(const std::string& name,  int hp, int mana, int dmg) 
    : SpellCaster(name, hp, mana, dmg, new SoldierState(this))
    , warrior(WarriorID::NECROMANCER), magician(MagicianID::SORCERER) {}

Necromancer::~Necromancer() {}

WarriorID Necromancer::getWarriorID() const {
    return this->warrior;
}

void Necromancer::setWarriorID(WarriorID newID) {
    this->warrior = newID;
}

MagicianID Necromancer::getMagicianID() const {
    return this->magician;
}

void Necromancer::update(Unit* enemy) {
    if ( enemy->getHealth().getHitPoints() == 0 ) {
        this->getHealth().addHitPoints(enemy->getHealth().getHitPointsLimit()/5);
        enemy->detach();
    }
}

void Necromancer::attack(Unit* enemy) {
    this->getAttack().attack(enemy);
    enemy->attach(this);
}

void Necromancer::counterAttack(Unit* enemy) {
    this->getAttack().counterAttack(enemy);
    enemy->attach(this);
}

void Necromancer::magicAttack(Unit* enemy, SpellID spell) {
    if ( spell == FREEZE || spell == HELLFIRE || spell == LIGHTNING || spell == ACIDMIST ) { 
        this->book->getSpellBook().find(spell)->second->mAttack(enemy, this);
    }
    enemy->attach(this);
}

File: Army(java)/Warriors/Necromancer.java
package Warriors;

import EnumID.*;
import Basic.*;
import Magic.*;

public class Necromancer extends SpellCaster implements IObserver {
    private WarriorID warrior;
    private MagicianID magician;
    
    public Necromancer(String name,  int hp, int mana, int dmg) {
        super(name, hp, mana, dmg);
        this.warrior = WarriorID.NECROMANCER;
        this.magician = MagicianID.SORCERER;
    }

    public WarriorID getWarriorID() {
        return warrior;
    }

    public void setWarriorID(WarriorID newID) {
        warrior = newID;
    }

    public MagicianID getMagicianID() {
        return magician;
    }
        
    public void update(Unit enemy) throws UnitIsDeadException {
        if ( enemy.getHealth().getHitPoints() == 0 ) {
            this.getHealth().addHitPoints(enemy.getHealth().getHitPointsLimit()/5);
            enemy.detach();
        }
    }

    public void attack(Unit enemy) throws UnitIsDeadException  {
        this.getAttack().attack(enemy);
        enemy.attach(this);
    }

    public void counterAttack(Unit enemy) throws UnitIsDeadException  {
        this.getAttack().counterAttack(enemy);
        enemy.attach(this);
    }

    public void magicAttack(Unit enemy, String spell) throws UnitIsDeadException, MagicManaOffException {
        if ( spell == "Freeze" || spell == "HellFire" || spell == "Lightning" || spell == "AcidMist" ) {
            book.getSpellBook().get(spell).mAttack(enemy, this);
        }
        enemy.attach(this);
    }
    
}